leader-election-impl/src/main/java/com/palantir/paxos/SplittingPaxosStateLog.java

```diff
@@ -0,0 +1,122 @@
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.immutables.value.Value;
+
+import com.palantir.common.persist.Persistable;
+import com.palantir.logsafe.Preconditions;
+
+/**
+ * This implementation of {@link PaxosStateLog} delegates all reads and writes of rounds to one of two delegates, as
+ * determined by the cutoff point. If a read or write does occur prior to the cutoff point, i.e., to the legacy delegate
+ * we update the appropriate metric. Remaining methods are delegated only to the current delegate.
+ */
+public final class SplittingPaxosStateLog<V extends Persistable & Versionable> implements PaxosStateLog<V> {
+    private final PaxosStateLog<V> legacyLog;
+    private final PaxosStateLog<V> currentLog;
+    private final Runnable markLegacyWrite;
+    private final Runnable markLegacyRead;
+    private final long cutoffInclusive;
+    private final AtomicLong leastLogEntry;
+
+    private SplittingPaxosStateLog(PaxosStateLog<V> legacyLog,
+            PaxosStateLog<V> currentLog,
+            Runnable markLegacyWrite,
+            Runnable markLegacyRead,
+            long cutoffInclusive,
+            AtomicLong leastLogEntry) {
+        this.legacyLog = legacyLog;
+        this.currentLog = currentLog;
+        this.markLegacyWrite = markLegacyWrite;
+        this.markLegacyRead = markLegacyRead;
+        this.cutoffInclusive = cutoffInclusive;
+        this.leastLogEntry = leastLogEntry;
+    }
+
+    public static <V extends Persistable & Versionable> PaxosStateLog<V> create(SplittingParameters<V> parameters) {
+        return new SplittingPaxosStateLog<>(
+                parameters.legacyLog(),
+                parameters.currentLog(),
+                parameters.markLegacyWrite(),
+                parameters.markLegacyRead(),
+                parameters.cutoffInclusive(),
+                new AtomicLong(Math.min(parameters.legacyLog().getLeastLogEntry(), parameters.cutoffInclusive())));
+    }
+
+    @Override
+    public void writeRound(long seq, V round) {
+        if (seq >= cutoffInclusive) {
+            currentLog.writeRound(seq, round);
+        } else {
+            markLegacyWrite.run();
+            legacyLog.writeRound(seq, round);
+            leastLogEntry.accumulateAndGet(seq, Math::min);
+        }
+    }
+
+    @Override
+    public byte[] readRound(long seq) throws IOException {
+        if (seq >= cutoffInclusive) {
+            return currentLog.readRound(seq);
+        } else {
+            markLegacyRead.run();
+            return legacyLog.readRound(seq);
+        }
+    }
+
+    @Override
+    public long getLeastLogEntry() {
+        return leastLogEntry.get();
+    }
+
+    @Override
+    public long getGreatestLogEntry() {
+        return currentLog.getGreatestLogEntry();
+    }
+
+    /**
+     * This implementation is a noop to ensure correctness of {@link #getLeastLogEntry()}.
+     */
+    @Override
+    public void truncate(long toDeleteInclusive) {
+        // noop
```

Comment:
`log.warn("Tried to truncate with an impl that doesn't support it")` or something like that please, this may confuse unsuspecting users.